File: crashreporter/src/main/java/net/iamsilver/crashreporter/adapter/ReportsAdapter.java
package net.iamsilver.crashreporter.adapter;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import net.iamsilver.crashreporter.R;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;

public class ReportsAdapter extends RecyclerView.Adapter<ReportsAdapter.ViewHolder> {

    private Context context;
    private OnItemClickListener onItemClickListener;
    private ArrayList<File> reports;

    //

    public ReportsAdapter(Context context, OnItemClickListener onItemClickListener) {
        this.context = context;
        this.onItemClickListener= onItemClickListener;
        reports = new ArrayList<>();
    }

    //

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
        View itemLayoutView = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.viewholder_reports_adapter, viewGroup, false);
        return new ViewHolder(itemLayoutView);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder viewHolder, int position) {
        final File file = reports.get(position);

        viewHolder.textViewReportTime.setText(file.getName().replaceAll("[a-zA-Z_.]", ""));
        viewHolder.textViewReportSummary.setText(readFirstLine(file));

        final int pos = position;
        viewHolder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (onItemClickListener != null) {
                    onItemClickListener.onClick(v, pos, file);
                }
            }
        });
    }

    @Override
    public int getItemCount() {
        return reports.size();
    }

    //

    public void setReports(ArrayList<File> reports) {
        this.reports.clear();
        this.reports.addAll(reports);
        notifyDataSetChanged();
    }

    //

    private static String readFirstLine(File file) {
        String line = "";
        try {
            BufferedReader reader = new BufferedReader(new FileReader(file));
            line = reader.readLine();
            reader.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return line;
    }

    //

    public interface OnItemClickListener {
        void onClick(View view, int position, File reportFile);
    }

    //

    class ViewHolder extends RecyclerView.ViewHolder {

        TextView textViewReportTime, textViewReportSummary;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);

            textViewReportTime = itemView.findViewById(R.id.textViewReportTime);
            textViewReportSummary = itemView.findViewById(R.id.textViewReportSummary);
        }

    }

}

File: crashreporter/src/main/java/net/cebular/crashreporter/internal/ReportsAdapter.kt
package net.cebular.crashreporter.internal

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import net.cebular.crashreporter.CrashReporter
import net.cebular.crashreporter.databinding.ViewholderReportsAdapterBinding
import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.IOException
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Locale

class ReportsAdapter(private val onItemClickListener: OnItemClickListener?) :
    RecyclerView.Adapter<ReportsAdapter.ViewHolder>() {

    fun interface OnItemClickListener {
        fun onClick(reportFile: File)
    }

    //

    class ViewHolder(private val binding: ViewholderReportsAdapterBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(file: File, onItemClickListener: OnItemClickListener?) {

            binding.textViewReportTime.text = formatFilenameDate(file.name)
            binding.textViewReportSummary.text = file.readFirstLine()

            binding.root.setOnClickListener {
                onItemClickListener?.onClick(file)
            }
        }

        fun File.readFirstLine(): String {
            var result = ""
            try {
                val reader = BufferedReader(FileReader(this))
                result = reader.readLine()
                reader.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
            return result
        }

        private fun formatFilenameDate(filename: String): String {
            if (filename.length < 15) {
                return filename
            }

            val date = filename.substring(0, 15)
            return try {
                val from = SimpleDateFormat(CrashReporter.FilenameTimestampDateFormatPattern, Locale.getDefault())
                val to = SimpleDateFormat("dd. MM. yyyy HH:mm:ss", Locale.getDefault())
                from.parse(date)?.let { to.format(it) } ?: date
            } catch (e: ParseException) {
                date
            }
        }
    }

    //

    private val reports: MutableList<File> = mutableListOf()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(ViewholderReportsAdapterBinding.inflate(LayoutInflater.from(parent.context), parent, false))
    }

    override fun getItemCount(): Int {
        return reports.size
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(reports[position], onItemClickListener)
    }

    fun setReports(newReports: List<File>) {
        val diff = DiffUtil.calculateDiff(DiffUtilCallback(reports, newReports))
        reports.clear()
        reports.addAll(newReports)
        diff.dispatchUpdatesTo(this)
    }

    private class DiffUtilCallback(val prev: List<File>, val curr: List<File>) : DiffUtil.Callback() {
        override fun getOldListSize(): Int = prev.size
        override fun getNewListSize(): Int = curr.size
        override fun areItemsTheSame(prevItemPosition: Int, currItemPosition: Int): Boolean =
            prev[prevItemPosition].name == curr[currItemPosition].name

        override fun areContentsTheSame(prevItemPosition: Int, currItemPosition: Int): Boolean = false
    }

}